src/main/scala/day2/Solution.scala

```diff
@@ -0,0 +1,49 @@
+package day2
+
+class Position(var x: Int, var y: Int) {
+  require(x >= 0 && x < 3)
+  require(y >= 0 && x < 3)
+
+  override def toString: String = s"Position($x, $y)"
+}
+
+object Solution {
+  private val Keys: Array[Array[Int]] = Array(
+    (1 to 3).toArray,
+    (4 to 6).toArray,
+    (7 to 9).toArray
+  )
+
+  private class Solution {
+    def run(instructions: List[List[Char]]): String = {
+      val code = new StringBuilder("")
+      val position = new Position(1, 1)
+
+      instructions.foreach(row => {
+        row.foreach {
```

Comment:
If you aim in a more functional approach, you should try to implement processing of `instructions` using recursion or map/reduce instead of these two `foreach` loops.